### ISSUE TITLE
writing to char crashes Chrome

### ISSUE THREAD

**Issue by aircable** [urish/ng-beacon-app#1]
Hi Urish,
I've been working on your code for a while and using it to async get notes from chars as well as writing to chars. I'm on Chrome (sometimes Canary) on OSX. Writing to char will eventually crash Chrome. It gets a seg fault.
Any idea how to avoid that?

**Comment by urish** [comment]
Hi @aircable,

Sounds like it is a bug in Chrome. Can you please open a bug report in their bug tracking system?

http://crbug.com/new

**Comment by aircable** [comment]
certainly, sorry for the inconvenience.
It happens when notifications are arriving at the same time when you write
to char.

I thought it may be something you have tried to work around by using the
BehaviorSubject.
It certainly looks to me that you have a 10ms delay between the calls:
.map(() => setTimeout(() => this.writableSubject.next(true), 10))

and then here:
).do(([rx, tx]) => {
setTimeout(() => this.writableSubject.next(true), 0);
});

Have you seen that issue when calling writechar too quickly?

Bug files for Chrome:
https://bugs.chromium.org/p/chromium/issues/detail?id=716655#c1

**Comment by urish** [comment]
Thanks! I mainly use Ubuntu (VM running inside Windows) for development, and it is pretty stable - I can't recall it crashing when notification was arriving at the same time when I wrote to char.

I can't recall the exact reason for the delay there. The second timeout was added in this commit:
https://github.com/urish/ng-beacon-app/commit/4d44a1425c1d3bc2d0677390f8fe02e89667f15f

As for the first run I am really not sure - judging by the date of [the commit](https://github.com/urish/ng-beacon-app/commit/8d1031e7063e6d0ae5b4f034aa1f790483acc750), it was the night before ngVikings conference, where I was giving a talk, so it could be that I just added it as a safety measure without much testing.

What kind of project are you working on?

**Comment by aircable** [comment]
And I thought, OSX is more stable.... well....
The guys at Chromium figured out why it was crashing. Now we need to hope
for a solution.
See update on the bug report.

I'm working on the Bluetooth Mesh protocol. I have some implementations
already
e.g.: https://github.com/aircable/SmartMesh
That is with Ionic 1, and it has the same problems. Can't write to a char
without crashing.
The app uses a mesh bridge (or proxy if you will) to get data into the
mesh. We make hardware
for that. See aircable.co for that.

Now I wanted to update to Ionic 3 and with that a new implementation with
TypeScript.
Your example came welcome. I copied a lot. But I would like to add error
handling for a real
usable app. Would you help? I have problems getting the debugging done and
catch errors
in a nice way (e.g. change screen while in discovery, reconnect when lost,
etc.)

**Comment by urish** [comment]
Wow, that was an amazingly quick turnaround for a bug fix!

Where can I learn more about the Bluetooth Mesh protocol? I looked up in your website, but couldn't find any technical documentation regarding the protocol itself. It is something open or proprietary?

As for error handling - it is basically just catching the rejections from any promises + adding an event listener for `"gattserverdisconnected"`. The exact action you will take depends on your application logic - but probably at least notify the user that there was some failure and perhaps an option to reconnect.

p.s. thanks for reporting the issue and working with the chrome guys to find the root cause - you have just helped to make the web platform even better :)

**Comment by aircable** [comment]
Indeed, absolutely amazing quick bug fix.
I will test that now. Maybe that just was it.....

You can read about the Bluetooth mesh (CSR style) here:
https://aircable.co/blog/aircable-news-1/post/bluetooth-smartmesh-8

Also here:
https://theiotmagazine.com/asset-tracking-in-the-bluetooth-smart-mesh-fc5a864e1e23
There are a bunch of links at the end for further reading.

Everything else is in the source code: https://github.com/aircable/SmartMesh

The Bluetooth SIG is in process standardizing the SIG Mesh. I have a huge
problem with that, since it is
so complex that the implementation does not fit in a standard 64k BLE chip.
You need 32bit processor at
100MHz or more and like 512k memory. My opinion is that it is blown out of
proportions regarding security.
And everything is mandatory making it actually pretty difficult to use. I
have the PS version of the standard
if you are interested.

Cheers
Juergen

**Comment by aircable** [comment]
you happen to know when a certain bug gets into a build version?
how to track that?

**Comment by urish** [comment]
Perhaps @beaufortfrancois can shed some light on the exact process :)

**Comment by beaufortfrancois** [comment]
According to https://bugs.chromium.org/p/chromium/issues/detail?id=716655#c7, I see that https://chromium.googlesource.com/chromium/src.git/+/6ce865fc28199a11109d3b72ad3b5eb52b3ae826/chrome/VERSION is:
```
MAJOR=60
MINOR=0
BUILD=3087
PATCH=0
```

TL;DR: This is in Chrome Canary

**Comment by aircable** [comment]
HI Uri,
I can't get the syntax right for adding an event listener. Can you help me
please?

return this.ble
    // 1: trigger discovery process in the browser
    .discover$( scanfilter )
    // 2: get that service
    .switchMap(( gatt: BluetoothRemoteGATTServer ) => {
        console.log("got gatt");
        this.gatt = gatt;
        this.ble.getDevice$().addEventListener('ongattserverdisconnected',
() => {
            console.log("disconnect event");
        });
        return this.ble.getPrimaryService$( gatt, BleUartService.UART_SERVICE );
    })
    // 3: now get the characteristics
    .switchMap( primaryService => this.connectRxTx( primaryService ));

**Comment by urish** [comment]
Do you mean the `ongattserverdisconnected` event?
You can use `gatt.device.addEventListener()` then

**Comment by beaufortfrancois** [comment]
I think I've spot the error:
```diff
- this.ble.getDevice$().addEventListener('ongattserverdisconnected',
+ this.ble.getDevice$().addEventListener('gattserverdisconnected',
```

**Comment by aircable** [comment]
Hi Uri,

Thanks for your reply. I've been watching you working on the ng-beacon-app
and it got it working
on my environment, transferring data both ways. Thanks for that!

Of course, I do a few things differently with our UART product. But I also
had to do some changes even beyond that.

Before I publish, I need to detach the disconnect event in the start
notification procedure.
So far, it has worked correctly, no matter which side disconnects, printing
the "disconnect" message.

But I need to show that on the UI. An event there would be great, but of
course, the event must happen
in the notification function too, to stop the notifications.
I need that event in your app-component.ts to set the "connected" boolean
correctly and reset everything.
Can you help? Create a new observable? Or inject another event into
this.uartData after the takeUntil?

Not idea yet how to do that. In your ble-uart-services.ts I do that:
Cheers
Juergen


this.uartData = Observable.fromEvent( this.uartChar,
'characteristicvaluechanged' )
    .map( value => String.fromCharCode.apply( null, new Uint8Array(
this.uartChar.value.buffer )))
    .takeUntil( Observable.fromEvent( this.gatt.device,
'gattserverdisconnected' ))
    .finally(() => {
        console.log( 'disconnected ');
        return this.uartChar.stopNotifications()
    })
    .share();
//
await this.uartChar.startNotifications();

**Comment by urish** [comment]
If I understand your question correctly, you want to know the uartData stream has closed (due to disconnection).

You can use the 3rd argument of subscribe:

```
this.uartData.subscribe(null, null, () => {
  // this will handle device disconnection
});
```

You can learn more about the arguments for subscribe in [rxjs docs](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/subscribe.md)

**Comment by aircable** [comment]
I could swear I did that before and it didn't work.
But, now, it just does! Magically ;-)
Thanks
Juergen
